### DISCUSSION TITLE
[Feature] Enter address or coordinates for media geolocation

### DISCUSSION THREAD

**Discussion by bpmcgill** [immich-app/immich#8551] [Feature Request]
### The feature

When I am editing the meta data of my images, I would like to be able to enter an address or coordinates for the geolocation of the media file. I have a lot of photos that were taken with the GPS information turned off.

Here is a photo of where the feature would be added
![image](https://github.com/immich-app/immich/assets/9805233/60ca2240-f4fe-4f43-8179-0f9d759c81af)

### Platform

- [X] Server
- [ ] Web
- [ ] Mobile

**Comment by bo0tzz** [comment]
Closing as a duplicate of #8380, which requests coordinate support. Supporting the detail level of addresses is not feasible without calling out to external APIs, which we won't do.

**Reply by bpmcgill** [reply]
That seems like a missed opportunity. Perhaps you should allow the user to make that decision and take on the financial responsibility of choosing the option. What API is used for this?

**Reply by JuGuSm** [reply]
> Closing as a duplicate of #8380, which requests coordinate support. Supporting the detail level of addresses is not feasible without calling out to external APIs, which we won't do.

That's strange because at the end the coordinates are stored in a xmp file so we could simply store directly the coordinates entered by the user.

**Reply by nielstron** [reply]
I agree with @bpmcgill that the user should have the option to call an external API if they so wish (configurable through the settings).

There are both non-free (i.e. [Google](https://developers.google.com/maps/documentation/geocoding/overview)) and OS-based (i.e. [Nominatim](https://nominatim.openstreetmap.org/ui/search.html)) options.

**Comment by nielstron** [comment]
A thread for a possible remedy for this issue.

it seems like the meat of the stuff is happening here:
- call the searchPlaces API from the frontend: https://github.com/immich-app/immich/blob/14e6d23eebe1687eab0cf8d079a66ad7d8040656/web/src/lib/components/shared-components/change-location.svelte#L74
- implementation of searchPlaces API by SQL query inside server stored locations: https://github.com/immich-app/immich/blob/14e6d23eebe1687eab0cf8d079a66ad7d8040656/server/src/repositories/search.repository.ts#L274
- searchPlaces service calls the searchPlace SQL query implementation: https://github.com/immich-app/immich/blob/14e6d23eebe1687eab0cf8d079a66ad7d8040656/server/src/services/search.service.ts#L49

I will try to replace this searchPlaces implementation with a (wrapped) call to the nominatim API. in a next step the user should be able to opt-in to this implementation.

**Reply by zackpollard** [reply]
Hey, appreciate the want to implement this, however we have discussed this numerous times internally and we are trying to remove calls to third-party external services, not add them. Also, we have specifically moved away from nominatim and OSM at our own expense (hosting a tile server) as we don't want the relatively large user base of immich adding additional load to their servers, which are paid for by donations and have very severe rate limits due to that. Check their usage policy here https://operations.osmfoundation.org/policies/nominatim/
We have improvements coming to the reverse geocoding implementation soon that will also bring improvements to search. We may also look into providing our own hosted search for this in the future, which might have compatiblity with the nominatim API itself, but for now we aren't looking to make changes in this area.

**Reply by nielstron** [reply]
Hi, thanks for the quick response. I am definitely not trying to overwhelm Nominatim or make it the default for all users, I was more looking into using an API in general that can then be swapped out by end users and potentially point to a hosted version of nominatim. I think it would be great if there is a hosted version of Nominatim by immich, though I do also see that the resource requirements would be quite significant. I also looked into paid alternatives, most cost as much as a server for self-hosting (and don't offer pay-by-usage unfortunately).

Overall though I think this particular feature would be well suited for using an external API because there are relatively few, user-tiggered, requests, that require significant resources to handle und thus forbid self-hosting for most end-users.

I will wait for further action until you pushed the new reverse geocoding features into the main database.

**Reply by bo0tzz** [reply]
>  I am definitely not trying to overwhelm Nominatim or make it the default for all users, I was more looking into using an API in general that can then be swapped out by end users and potentially point to a hosted version of nominatim.

This is sensible, but realistically most users would end up just pointing at the freely available OSM instance anyways.

**Reply by kubikaugustyn** [reply]
Hi, I'm also looking for a way to do this, as the reverse geocoding gives me a totally different town for my house's position. I'd like to use the Czech Mapy.com reverse geocoding API, same as I changed my map to render their tiles. My family likes Mapy.com's maps more than OpenStreetMaps, so they're happy to have their tiles in the map.
However, as of now, you can only turn reverse geocoding on and off in the server settings, not use a custom server URL. I get that most people `would end up just pointing at the freely available OSM instance anyways`. But maybe if you added a disclaimer explaining that you should only add an API if you're paying for an API key or something and if the user tried to add the public OSM server, you'd also show a disclaimer, maybe, just maybe it could be possible?
I'd be happy with any API format, I can convert the data myself.
Also, as https://github.com/immich-app/immich/discussions/22688 mentions, it could be an environment variable to "hide" it from normal users? For example: `IMMICH_REVERSE_GEOCODING_URL=https://geo.example.com/reverse?lat={lat}&lon={lon}`

Best regards,
Jakub Augustýn

**Comment by nielstron** [comment]
Re-animating this discussion in a different direction: I spent today building a completely local lookup tool for streets, sights, airports and similar landmarks. It all takes <1GB disk space and runs client-side, e.g., should work even if immich is hosted on a potato.

An example for the lookup functionality is hosted on a Github page (no backend used at all): https://nielstron.github.io/local-street-db/

I'd be more than happy if this finds its way into the immich ecosystem and can help integrating it.

**Reply by bo0tzz** [reply]
@zackpollard :eyes:

**Comment by yves-chevallier** [comment]
Landing here from #30134 (closed as dup, fair enough). Adding the technical angle in case it helps: the reason address/POI search returns nothing today is structural, search.repository.ts only queries geodata_places, which is populated purely from the GeoNames cities500 set (cities ≥ 500 pop), so no POI, business, or address is ever in the data.

Rather than a one-off address lookup, it might be worth a small GeocodingProvider seam so the backend is swappable: keep the local geodata_places search as the offline default, and let admins optionally point at a self-hosted Nominatim/Photon (or Google/Mapbox) for fuzzy POI search. Results normalize to the existing PlacesResponseDto, so the UI barely changes. Bonus: that same config could later back the reverse-geocoding requests in #22688 / #18198. Happy to help scope it if there's appetite.
